Sources/Workspace/Workspace.swift

```diff
@@ -289,88 +297,91 @@ public class Workspace {
-                mirrors: mirrors
-            )
-        }
-
-        self.additionalFileRules = additionalFileRules
-        self.resolverUpdateEnabled = resolverUpdateEnabled
-        self.resolverPrefetchingEnabled = resolverPrefetchingEnabled
-        self.resolverFingerprintCheckingMode = resolverFingerprintCheckingMode
+        try self.init(
+            fileSystem: fileSystem,
+            location: location,
+            registries: registries,
+            authorizationProvider: authorizationProvider,
+            fingerprints: customFingerprintStorage,
+            mirrors: mirrors,
+            resolutionConfiguration: resolutionConfiguration,
+            customToolsVersion: customToolsVersion,
+            customManifestLoader: customManifestLoader,
+            customRepositoryManager: customRepositoryManager,
+            customRepositoryProvider: customRepositoryProvider,
+            customRegistryClient: customRegistryClient,
+            customIdentityResolver: customIdentityResolver,
+            customHTTPClient: customHTTPClient,
+            customArchiver: customArchiver,
+            customChecksumAlgorithm: customChecksumAlgorithm,
+            delegate: delegate
+        )
+    }
 
-        self.state = WorkspaceState(dataPath: self.location.workingDirectory, fileSystem: fileSystem)
+    /// Create a new package workspace.
+    ///
+    /// This initializer is designed for use cases when the workspace needs to be highly customized such as testing.
+    /// In other cases, use the other, more straight forward, initializers
+    ///
+    /// This will automatically load the persisted state for the package, if
+    /// present. If the state isn't present then a default state will be
+    /// constructed.
+    ///
+    /// - Parameters:
+    ///   - fileSystem: The file system to use.
+    ///   - location: Workspace location configuration.
+    ///   - registries: Configuration for registries
+    ///   - authorizationProvider: Provider of authentication information for outbound network requests.
+    ///   - fingerprints: Dependencies fingerprints.
+    ///   - mirrors: Dependencies mirrors (overrides).
+    ///   - resolutionConfiguration: Configuration to fine tune the dependency resolution behavior.
+    ///   - customManifestLoader: A custom manifest loader.
+    ///   - customRepositoryProvider: A custom repository provider.
+    ///   - delegate: Delegate for workspace events
+    public convenience init(
+        fileSystem: FileSystem,
+        location: Location,
+        registries: RegistryConfiguration? = .none,
+        authorizationProvider: AuthorizationProvider? = .none,
+        fingerprints: PackageFingerprintStorage? = .none,
+        mirrors: DependencyMirrors? = .none,
+        resolutionConfiguration: ResolutionConfiguration? = .none,
+        // optional customization, primarily designed for testing
+        // FIXME: TBD what is actually needed here by libSwiftPM consumers
```

Comment:
@abertelrud @neonichu need you input here

Reply:
Just so I understand: code that uses libSwiftPM can always use the full initializer, right, so this is just a question about what's in the convenience initializer?

If so then this seems like a good set for simple users of libSwiftPM.  Looks like anyone who wants custom locations for, say, the mirrors, would still need to use the full version?

It's still useful to have a minimal version I think.

Reply:
@abertelrud I am trying to rework what that "full initializer" looks like, to separate the concerns of testing (which need allot of customization options) and libSwiftPM consumers (which need some customization options). For example, I dont think libSwiftPM consumers should be able to override the checksum algorithm, but we do need that for testing.

IOW the question is what out of the customization options that we offer is actually needed by  libSwiftPM consumers and _must_ be part of the public initializer. 

Here is the customization options we offer today and we need to decide what we must keep:

```
// optional customization, primarily designed for testing but also used in some cases by libSwiftPM consumers
customRegistriesConfiguration: RegistryConfiguration? = .none,
customFingerprints: PackageFingerprintStorage? = .none,
customMirrors: DependencyMirrors? = .none,
customToolsVersion: ToolsVersion? = .none,
customManifestLoader: ManifestLoaderProtocol? = .none,
customRepositoryManager: RepositoryManager? = .none,
customRepositoryProvider: RepositoryProvider? = .none,
customRegistryClient: RegistryClient? = .none,
customIdentityResolver: IdentityResolver? = .none,
customHTTPClient: HTTPClient? = .none,
customArchiver: Archiver? = .none,
customChecksumAlgorithm: HashAlgorithm? = .none,
```

Reply:
I think mirrors, manifest loader, repository manager/provider and container provider is the minimum we should expose. 

I could see the HTTP client being useful as well, if someone tries to bring SwiftPM to a platform that has strict rules and conventions about how networking is supposed to work.

The rest seem mostly be useful for testing to me.

Reply:
I agree with Boris here:  mirrors, repository provider, and manifest loader are the most important.  Not sure about why a custom repository manager would be needed, but provider is one that client code definitely needs to customize.